packages/cli/src/publish.test.ts

```diff
@@ -0,0 +1,243 @@
+  const otherPreset = 'other-preset';
+  let signer: ethers.Signer;
+  const deployDataLocalFileName = `${packageName.replace(':', '_')}_${chainId}-${preset}.txt`;
+  const miscData = { misc: 'info' };
+  const metaData = { itsMeta: 'data' };
+
+  const testPkgData: DeploymentInfo = {
+    def: { name: 'package', version: '1.2.3', provision: { dummyStep: { tags } } } as any,
+    state: {
+      'provision.dummyStep': {
+        hash: '',
+        version: 1,
+        artifacts: {},
+      },
+    },
+    status: 'complete',
+    miscUrl: 'file:/usecannon.com/misc',
+    meta: { itsMeta: 'data' },
+    options: {},
+  };
+
+  const testPkgDataIpfsUrl = 'ipfs:/test-ipfs-url';
+  const testPkgDataNewIpfsUrl = 'ipfs:/test-ipfs-new-url';
+  const testPkgMetaIpfsUrl = 'ipfs:/test-ipfs-meta-url';
+  const testPkgMiscIpfsUrl = 'ipfs:/test-ipfs-misc-url';
+  const testPkgNewMetaIpfsUrl = 'ipfs:/test-ipfs-new-meta-url';
+
+  it.skip('should upload the package to IPFS', async () => {
+    // TODO - this functionality will be moved to a new command `extract-link`
+    // https://github.com/usecannon/cannon/issues/280
+  });
+
+  beforeAll(async () => {
+    jest.resetAllMocks();
+
+    jest.spyOn(settings, 'resolveCliSettings').mockImplementation(
+      jest.fn().mockReturnValue({
+        ipfsUrl: 'http://127.0.0.1:5001',
+        registryProviderUrl: 'http://localhost:3000',
+        registryAddress: ethers.constants.AddressZero,
+        registryChainId: '123', // or whatever value is appropriate in your case
+        cannonDirectory: '/cannon/directory/',
+        // Add other properties as needed
+      })
+    );
+
+    signer = ethers.Wallet.createRandom().connect(
+      new CannonWrapperGenericProvider({}, new ethers.providers.JsonRpcProvider())
+    );
+  });
+
+  beforeEach(() => {
+    // Reset mocks
+    jest.clearAllMocks();
+
+    // @ts-ignore
+    jest.spyOn(fs, 'readdir').mockResolvedValue([deployDataLocalFileName, deployDataLocalFileName + '.meta']);
+
+    const cliSettings = resolveCliSettings();
+
+    jest.spyOn(LocalLoader.prototype, 'read').mockImplementation(async (url) => {
+      switch (url) {
+        case 'file:/usecannon.com/misc':
+          return miscData;
+        case 'file:/usecannon.com/meta':
+          return metaData;
+        case 'file:/usecannon.com':
+        case url:
+          return testPkgData;
+      }
+      return '';
+    });
+
+    jest.spyOn(fs, 'readFile').mockImplementation((_path) => {
+      const testPkgDataFilePath = path.join(cliSettings.cannonDirectory, 'tags', deployDataLocalFileName);
+      switch (_path) {
+        case testPkgDataFilePath:
+          return Promise.resolve(Buffer.from(testPkgDataIpfsUrl));
+        case testPkgDataFilePath + '.meta':
+          return Promise.resolve(Buffer.from(testPkgMetaIpfsUrl));
+      }
+      return Promise.resolve(Buffer.from(''));
+    });
+    jest.spyOn(IPFSLoader.prototype, 'read').mockImplementation((url: string): Promise<any> => {
```

Comment:
Instead of using mocks you could have probably used the in memory registry and loader

Reply:
@dbeal-eth I couldn't find the `InMemoryLoader` class anywhere in the code, are you sure of the name?
Also, particularly in this case, we don't have control over the type of loader that is instantiated as I see. The `publish` method instantiates one itself, do you mean I mock the whole OnChainRegistery class?

Reply:
As I said we don't have control over the type of loader created, or the registry. So I merge this PR.